conductor/__beta__/job/job.py

```diff
@@ -0,0 +1,144 @@
+        return conductor.lib.package_utils.merge_package_environments(packages, base_env=self.environment)
+    
+    def get_output_path(self):
+        return self.output_path
+    
+    def scan_for_dependencies(self):
+        return []
+    
+    def get_dependencies(self):
+        
+        if self._dependencies is None and self._dependency_scan_enabled:            
+            self._dependencies = self.scan_for_dependencies()
+            
+        return self._dependencies + self.upload_paths
+
+    def submit_job(self):
+        
+        self.validate_job()
+        
+        data = { "upload_paths": self.get_dependencies(),
+                 "software_package_ids": self.software_packages_ids, 
+                 "tasks_data": self._get_task_data(), 
+                 "owner": self.owner, 
+                 "frame_range": self._get_frame_range(),
+                 "environment": self._get_environment(), 
+                 "priority": self.priority,
+                 "location": self.location, 
+                 "instance_type": self.instance_type, 
+                 "preemptible": self.preemptible, 
+                 "metadata": self.metadata, 
+                 "local_upload": self.local_upload, 
+                 "autoretry_policy": self.auto_retry_policy,
+                 "chunk_size": self.chunk_size, 
+                 "project": self.project,
+                 "output_path": self.get_output_path(), 
+                 "job_title": self.job_title}
+        
+        if self.docker_image:
+            data["docker_image"] = self.docker_image
+        
+        for key, value in os.environ.items():
+            if key.startswith("CONDUCTOR_JOBPARM_"):
+                job_parm_key = key.replace("CONDUCTOR_JOBPARM_", "")
+                
+                if value.lower() == "true":
+                    value = True
+                
+                elif value.lower() == "false":
+                    value = False
+                    
+                try:
+                    value = int(value)
+                except ValueError:
+                    pass
+                    
+                data[job_parm_key.lower()] = value
+        
+        LOG.debug("Job Parameters:")
+        for k, v in data.iteritems():
+            LOG.debug("  {}: '{}'".format(k, v))
+        
+        submitter = conductor_submit.Submit(data)
+        
+        response, response_code = submitter.main()
+        LOG.debug("Response Code: %s", response_code)
+        LOG.debug("Response: %s", response)
+         
+        if response_code in [201, 204]:
+            LOG.info("Submission Complete")
+ 
+        else:
+            LOG.error("Submission Failure. Response code: %s", response_code)
+            raise JobError("Submission Failure. Response code: %s", response_code)
+ 
+        self.conductor_job_id = response['jobid']
+         
+        return self.conductor_job_id
+    
+    @classmethod
+    def get_klass(cls, cmd):

```

Comment:
This method doesn't appear to be used anywhere...which is not automatically an issue, but it does remind me that one of the objectives of this api development process is _seeing_ how the api would be used from a consumer's perspective.   This method doesn't quite smell right to me, and I'm wondering if  an example usage would further validate (or dismiss) that concern.

The reason that I generally take issue with this method is that that it's deriving information (i.e. making an assumption) in order to drive logic.  Submitting a job  should be a declarative process.  The person/code submitting the job needs to construct the command to execute, the environment required, the files and software required, etc.  Therefore there shouldn't be a need for a method that "guesses" as to what job class should be used (this is information that should be known from the start, i.e _im using maya_).  Parsing a command is not a reliable means to derive this information (what happens if the command is simply `bash ~/myscript.sh`...which may subsequently end up executing maya....or katana, etc).

With all of that said, I'm pretty sure why you're doing this: Conductor is serving as a dummy pass through for Deadline needs (therefore the conductor job class may not be privy to the information/logic that was used to describe the customer's deadline job).  If that's the case, I think this is going to be an ongoing hurdle, and we'll have to cross some of these bridges when we get there.  But at the very least, I don't think this method should live on this class. I would define it within a context that is specific to deadline. Whether that's within a method on the deadline class, or as a standalone function which clearly spells out what it's doing, and why it must be done that way.

If we _do_ end up wanting to declare a class method that returns the appropriate class (given a specific piece of data), I would think that specific piece of data should be one of our own constructs: either a _package_ **id**,  or a _package_ **product**

Reply:
You're mostly right - it's used by Tractor.

Initially, all this stuff was written to by a POC and sample code for Tractor. When I started the Deadline plugin I re-used it.

My gut feeling is that it's OK for class/static methods to not be very general and they could be special purpose. It's maybe not ideally placed here but I find more often than not, there's no better place. I'm definitely open to a better solution I just haven't come across one yet, so this is the paradigm I tend to follow.